config/environments/production.rb

```diff
@@ -130,13 +130,12 @@
   config.action_mailer.perform_deliveries = true
   config.action_mailer.default_url_options = { host: protocol + ENV["APP_DOMAIN"].to_s }
   ActionMailer::Base.smtp_settings = {
-    address: "smtp.sendgrid.net",
-    port: "587",
-    authentication: :plain,
-    user_name: "apikey",
-    password: ENV["SENDGRID_API_KEY"],
-    domain: ENV["APP_DOMAIN"],
-    enable_starttls_auto: true
+    address: ENV["SMTP_ADDRESS"].presence || "smtp.sendgrid.net",
+    port: ENV["SMTP_PORT"].presence || 587,
+    authentication: ENV["SMTP_AUTHENTICATION"].presence || :plain,
+    user_name: ENV["SMTP_USER_NAME"].presence || "apikey",
+    password: ENV["SMTP_PASSWORD"].presence || ENV["SENDGRID_API_KEY"],
```

Comment:
[nonblocking] I would much rather we adjust the env vars to avoid vendor-specific environment variables, _however_ what you've got seems to allow for migration without breaking anyone else's installation.

The reason this touches a nerve is the `REDISCLOUD_URL` that overrode `REDIS_URL` for us last week or the week before, but I think what you're doing here does that better. When we get versioning out the door, this is the kind of thing that will need a changelog entry like this:

> Deprecated `SENDGRID_API_KEY` in favor of `SMTP_PASSWORD`. The deprecated environment variable will be removed in version `stable.2021XXYY.0`.

Would love to hear your thoughts on this braindump 🙂

Reply:
I previously didn't add this fallback and was planning to change all env var everywhere before emerging this PR. I changed my mind because this does provide a smoother transition and additional change needed from System's side can be won't block this PR from being merged.

Reply:
**Non-blocking**: for clarity's sake it would be easier if there were two independent "old, sendgrid style" config and a new "smtp is configurable" style options for smtp_settings - if this is all scaffolding while we get the config/env vars in deployed systems, and it will disappear soon after that's done, no objections.  Maybe "if ENV has key SENDGRID_API_KEY - legacy setup - else - new setup", or the reverse is what I meant.

The closer we get to self-hosted open source deploys being something to factor in, the harder it will be to reason about when it's okay to remove old options and we _might_ live with this code for longer than you expected.

The reason I'm suggesting two blocks is while it feels like we have only two alternatives (SMTP config was explicit OR we use legacy sendgrid config) it looks like we have 5 alternatives and they _could_ be independent. I don't want them to be independent (either we're all in column A or we're all in column B and no mixing).

Reply:
Thank you for that amazing suggestion @djuber ! I updated it.